packages/cw721/src/testing/contract_tests.rs

```diff
@@ -0,0 +1,1547 @@
+
+    // unknown nft returns error
+    let _ = contract
+        .query_nft_info(deps.as_ref(), env.clone(), "unknown".to_string())
+        .unwrap_err();
+
+    // this nft info is correct
+    let info = contract
+        .query_nft_info(deps.as_ref(), env.clone(), token_id.clone())
+        .unwrap();
+    assert_eq!(
+        info,
+        NftInfoResponse::<DefaultOptionMetadataExtension> {
+            token_uri: Some(token_uri),
+            extension: None,
+        }
+    );
+
+    // owner info is correct
+    let owner = contract
+        .query_owner_of(deps.as_ref(), mock_env(), token_id.clone(), true)
+        .unwrap();
+    assert_eq!(
+        owner,
+        OwnerOfResponse {
+            owner: String::from("medusa"),
+            approvals: vec![],
+        }
+    );
+
+    // Cannot mint same token_id again
+    let mint_msg2 = Cw721ExecuteMsg::Mint {
+        token_id: token_id.clone(),
+        owner: String::from("hercules"),
+        token_uri: None,
+        extension: None,
+    };
+
+    let allowed = mock_info(MINTER_ADDR, &[]);
+    let err = contract
+        .execute(deps.as_mut(), mock_env(), allowed, mint_msg2)
+        .unwrap_err();
+    assert_eq!(err, Cw721ContractError::Claimed {});
+
+    // list the token_ids
+    let tokens = contract
+        .query_all_tokens(deps.as_ref(), env, None, None)
+        .unwrap();
+    assert_eq!(1, tokens.tokens.len());
+    assert_eq!(vec![token_id], tokens.tokens);
+}
+
+#[test]
+fn mint_with_metadata() {
+    // case 1: mint with valid metadata
+    {
+        let mut deps = mock_dependencies();
+        let contract = setup_contract(deps.as_mut());
+
+        let token_id = "1".to_string();
+        let token_uri = "ipfs://foo.bar".to_string();
+        let mint_msg = Cw721ExecuteMsg::Mint {
+            token_id: token_id.clone(),
+            owner: String::from("medusa"),
+            token_uri: Some(token_uri.clone()),
+            extension: Some(Metadata {
+                image: Some("ipfs://foo.bar/image.png".to_string()),
+                image_data: Some("image data".to_string()),
+                external_url: Some("https://github.com".to_string()),
+                description: Some("description".to_string()),
+                name: Some("name".to_string()),
+                attributes: Some(vec![Trait {
+                    trait_type: "trait_type".to_string(),
+                    value: "value".to_string(),
+                    display_type: Some("display_type".to_string()),
+                }]),
+                background_color: Some("background_color".to_string()),
+                animation_url: Some("ssl://animation_url".to_string()),
+                youtube_url: Some("file://youtube_url".to_string()),
+            }),
+        };
+
+        let info_minter = mock_info(MINTER_ADDR, &[]);
+        let env = mock_env();
+        contract
+            .execute(deps.as_mut(), env.clone(), info_minter, mint_msg)
+            .unwrap();
+    }
```

Comment:
happy path testing on contract level, where new traits `Update` and `Validate` are taken into account. So minting now checks new features and make sure e.g. URLs are correct, Strings are not empty - in case Some(string) is provided.